what4/src/What4/Utils/BVDomain.hs

```diff
@@ -1,18 +1,503 @@
+
+--------------------------------------------------------------------------------
+-- Shifts
+
+shl :: (1 <= w) => NatRepr w -> BVDomain w -> BVDomain w -> BVDomain w
+shl w _a _b = any w -- TODO
+
+lshr :: (1 <= w) => NatRepr w -> BVDomain w -> BVDomain w -> BVDomain w
+lshr w _a _b = any w -- TODO
+
+ashr :: (1 <= w) => NatRepr w -> BVDomain w -> BVDomain w -> BVDomain w
+ashr w _a _b = any w -- TODO
+
+--------------------------------------------------------------------------------
+-- Arithmetic
+
+add :: (1 <= w) => BVDomain w -> BVDomain w -> BVDomain w
+add a b =
+  case a of
+    BVDAny _ -> a
+    BVDInterval _ al aw ->
+      case b of
+        BVDAny _ -> b
+        BVDInterval mask bl bw ->
+          interval mask (al + bl) (aw + bw)
+
+negate :: (1 <= w) => BVDomain w -> BVDomain w
+negate a =
+  case a of
+    BVDAny _ -> a
+    BVDInterval mask al aw -> BVDInterval mask ((-ah) .&. mask) aw
+      where ah = al + aw
+
+scale :: (1 <= w) => Integer -> BVDomain w -> BVDomain w
+scale k a
+  | k == 0 = BVDInterval (bvdMask a) 0 0
+  | k == 1 = a
+  | otherwise =
+    case a of
+      BVDAny _ -> a
+      BVDInterval mask al aw
+        | k >= 0 -> interval mask (k * al) (k * aw)
+        | otherwise -> interval mask (k * ah) (k * aw)
+        where ah = al + aw
+
+mul :: (1 <= w) => BVDomain w -> BVDomain w -> BVDomain w
+mul a b
+  | isSingletonZero a = a
+  | isSingletonZero b = b
+  | isBVDAny a = a
+  | isBVDAny b = b
+  | otherwise = interval mask cl cw
+    where
+      (cl, ch) = mulRange (zbounds a) (zbounds b)
+      cw = ch - cl
+      mask = bvdMask a
+      size = mask + 1
+      zbounds (BVDAny m) = (0, m)
+      zbounds (BVDInterval _ l w) = (l', l' + w)
+        where l' = if 2*l + w >= size then l - size else l
+
+mulRange :: (Integer, Integer) -> (Integer, Integer) -> (Integer, Integer)
+mulRange (al, ah) (bl, bh) = (cl, ch)
+  where
+    (albl, albh) = scaleRange al (bl, bh)
+    (ahbl, ahbh) = scaleRange ah (bl, bh)
+    cl = min albl ahbl
+    ch = max albh ahbh
+
+scaleRange :: Integer -> (Integer, Integer) -> (Integer, Integer)
+scaleRange k (lo, hi)
+  | k < 0 = (k * hi, k * lo)
+  | otherwise = (k * lo, k * hi)
+
+udiv :: (1 <= w) => BVDomain w -> BVDomain w -> BVDomain w
+udiv a b = interval mask ql (qh - ql)
+  where
+    mask = bvdMask a
+    (al, ah) = ubounds a
+    (bl, bh) = ubounds b
+    ql = al `div` max 1 bh -- assume that division by 0 does not happen
+    qh = ah `div` max 1 bl -- assume that division by 0 does not happen
+
+urem :: (1 <= w) => BVDomain w -> BVDomain w -> BVDomain w
+urem a b
+  | qh == ql = interval mask rl (rh - rl)
+  | otherwise = interval mask 0 (bh - 1)
+  where
+    mask = bvdMask a
+    (al, ah) = ubounds a
+    (bl, bh) = ubounds b
+    (ql, rl) = al `divMod` max 1 bh -- assume that division by 0 does not happen
+    (qh, rh) = ah `divMod` max 1 bl -- assume that division by 0 does not happen
+
+sdiv :: (1 <= w) => BVDomain w -> BVDomain w -> BVDomain w
```

Comment:
Are the signed divisions fundamentally more difficult?

Reply:
The main issue is that I need to know for sure how `sdiv` and `srem` are supposed to behave in both what4 and Cryptol, so that I can verify the implementation. And it seems that there may be some bugs around signed division in Cryptol at the moment...

Reply:
Yeah, this stuff is tricky.  The what4 documentation claims to round toward 0, but looking at the SMTLib2 definition of `bvsdiv`, I'm not entirely certain what it is supposed to do.

Reply:
Given that `bvudiv` rounds to zero, the `bvsdiv` in SMTLIB also always rounds to zero (http://smtlib.cs.uiowa.edu/logics-all.shtml#QF_BV)

Reply:
Ok, I think I've convinced myself that `bvsdiv` and `bvsrem` in SMTLIB match the `/$` and `%$` operations that we have in cryptol. I'll go ahead and try to implement the abstract domain operations for `sdiv` and `srem` in cryptol and then port them to haskell here.

Reply:
Implemented in revisions 8eb6dd7ae726ddc4fbbe71be23c1782b5ba87bf1 and 946c224b6a3f52c080120aaa6ec82d688bbe6274